src/main/java/io/confluent/connect/jdbc/sink/metadata/FieldsMetadata.java

```diff
@@ -161,19 +160,29 @@ public static FieldsMetadata extract(
           }
           keyFieldNames.addAll(configuredPkFields);
         }
+        for (String fieldName : keyFieldNames) {
+          final Schema fieldSchema = valueSchema.field(fieldName).schema();
+          allFields.put(fieldName, new SinkRecordField(fieldSchema.type(), fieldName, true, false, fieldSchema.defaultValue()));
+        }
       }
       break;
 
     }
 
+    final Set<String> nonKeyFieldNames = new LinkedHashSet<>();
     if (valueSchema != null) {
-      for (Field field : valueSchema.fields()) {
-        final boolean isKeyField = keyFieldNames.contains(field.name());
-        if (!isKeyField) {
-          nonKeyFieldNames.add(field.name());
+      for (Field field: valueSchema.fields()) {
+        if (keyFieldNames.contains(field.name())) {
+          continue;
+        }
+        if (!fieldsWhitelist.isEmpty() && !fieldsWhitelist.contains(field.name())) {
+          continue;
         }
+
+        nonKeyFieldNames.add(field.name());
+
         final Schema fieldSchema = field.schema();
-        allFields.put(field.name(), new SinkRecordField(fieldSchema.type(), field.name(), isKeyField, !isKeyField && fieldSchema.isOptional(), fieldSchema.defaultValue()));
+        allFields.put(field.name(), new SinkRecordField(fieldSchema.type(), field.name(), false, fieldSchema.isOptional(), fieldSchema.defaultValue()));
```

Comment:
`allFields` is also being manipulated in the `switch (pkMode)` block above.

Previously, for pkMode = KAFKA, RECORD_KEY we were adding to `allFields` in the above block, but for RECORD_VALUE we were only adding to `allFields` over here along with all the non-key fields. Now, the key fields for RECORD_VALUE pkMode are also inserted into `allFields` in the switch block. Over here we only deal with the non-key fields. https://github.com/confluentinc/kafka-connect-jdbc/pull/131/files#diff-4b72122fa7b6eb8a030b667da3073382R163